pkg/kubelet/util/pluginwatcher/plugin_watcher.go

```diff
@@ -0,0 +1,94 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pluginwatcher
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net"
+	"strings"
+	"time"
+
+	"github.com/fsnotify/fsnotify"
+	"google.golang.org/grpc"
+)
+
+// PluginInfo contains gRPC connection info
+type PluginInfo struct {
+	SockPath string
+}
+
+// WatchPlugins watches for the creation of plugin sockets at the path
+func WatchPlugins(path string, outChan chan PluginInfo) error {
+	files, err := ioutil.ReadDir(path)
+	if err != nil {
+		return err
+	}
+	watcher, err := fsnotify.NewWatcher()
```

Comment:
+1. Maybe take a look at the existing pkg/volume/flexvolume/probe.go and see whether we can follow a similar flow. Some example functionalities that the current pkg/volume/flexvolume/probe.go has and we may also consider:
- create the plugin directory if it doesn't exist
- recursively add watcher for sub directories
- watchEventInterval and watchEventLimit
- watch remove events

Reply:
It looks like that code lacks a stop function

Reply:
@verult Thanks a lot for your helpful inputs! Agree we should try to reuse FSWatcher interface. For now, I think we may leave remove events and recursively watch out of this PR. For possible future extension, I would suggest that PluginWatcher passes the absolute path in the channel, and the PluginWatcher caller can do its own naming convention verification.

For CSI, is it possible that we use /var/lib/kubelet/plugins/csi/[SanitizedCSIDriverName].sock naming convention to work with a flat directory? If we have to use the proposed /var/lib/kubelet/plugins/[SanitizedCSIDriverName]/csi.sock format, we can add recursive watch in PluginWatcher in a later PR.

For device plugin, I think the reason why we can't just pass resource name in Registration response is because the advertised resource name may fail Kubelet's validation, if e.g., it doesn't follow extended resource name convention, or it collides with an existing name. I am fine with the naming format currently used in this PR. Something we may consider is whether we want to allow socket name with some timestamp suffix to allow a new instance to come up without tearing down the old plugin instance to ensure zero downtime. But I think we can leave that part out for now, given that we expect most plugins to be deployed as DaemonSet containers whose upgrades require container restarts.

Reply:
So I thought about it more and I think a good solution could be to have to following model:
- The socket has the same format as the resource name (e.g: /var/lib/kubelet/plugins/device-plugins/nvidia.com/gpu.sock for the socket)
- The gRPC server advertises the same name as it's socket (e.g: "nvidia.com/gpu")
- Kubelet then validates that the socket name matches the advertised name

> or it collides with an existing name

That would be the same behavior as re-registration.

Reply:
I am also fine with this model. It is just that ProbeWatcher has to support recursive watching then, and I don't know whether that would create some performance concerns as inotify man page mentions that "Inotify monitoring of directories is not recursive: to monitor subdirectories under a directory, additional watches must be  created.   This  can take a significant amount time for large directory trees."

Reply:
> Inotify monitoring of directories is not recursive: to monitor subdirectories under a directory, additional watches must be created. This can take a significant amount time for large directory trees.

Pretty sure that's meant for [large amount of directories (>10000)](http://www.linuxjournal.com/article/8478) rather than our 3 device plugins.

Reply:
> Something we may consider is whether we want to allow socket name with some timestamp suffix to allow a new instance to come up without tearing down the old plugin instance to ensure zero downtime

We could just require the directory to match the vendor name instead of exact match of the resource name.
e.g: /var/lib/kubelet/plugins/device-plugins/nvidia.com/gpu-v1.8.sock and 
      /var/lib/kubelet/plugins/device-plugins/nvidia.com/gpu-v1.9.sock

Could both advertise resources under `nvidia.com/*`
Naming convention after `nvidia.com/` would then be left to the vendor.